Add notes for Build a Site from Scratch with Next.js, TypeScript, Emotion, and Netlify (#116)
eggheadio/eggheadio-course-notes: build-a-site-from-scratch-with-nextjs-typescript-emotion-and-netlify/notes/08-use-static-generation-with-getstaticprops-in-next.js-with-typescript.md

- Because the content on our blog is not likely to change often we're going to have the content of our site generate at build time, to do that we're going to use `getStaticProps`.
- `getStaticProps` is a function that a page, **and only a page** can run at build time in order to pre-render the page. The props that are returned from this function are going to be passed into the page component.
- For this example we're going to use [JSON Placeholder API](https://jsonplaceholder.typicode.com/posts) to get sample posts, but Next.js supports a variety of ways to get static content.
- Here's our sample code that fetches posts (❗️Note that this goes in the `index.tsx` file):

```js
export const getStaticProps = async () => {
  const res = await fetch("https://jsonplaceholder.typicode.com/posts");

  const posts: Post[] = await res.json();

  return {
    props: {
      posts,
    },
  };
};
```

- 🧩 To define a type in typescript use the following syntax:

```js
type Post = {
  userId: number,
  id: number,
  title: string,
  body: string,
};
```

> - When we try to pass these props in to our Posts function Typescript will throw an error, since it doesn't know what the type is of the static props. Next.js provides an `InferGetStaticPropsType` that helps Typescript infer the tyep of the static props.

Reviewer: ```suggestion
- When we try to pass these props into our Posts function Typescript will throw an error, since it doesn't know what the type is of the static props. Next.js provides an `InferGetStaticPropsType` that helps Typescript infer the type of the static props.
```
